src/dnmd/write.c

```diff
@@ -807,8 +793,115 @@ bool md_add_new_row_to_list(mdcursor_t list_owner, col_index_t list_col, mdcurso
+{
+    if (!col_points_to_list(&list_owner, list_col))
+       return false;
+
+    // Get the range of rows already in the parent's child list.
+    // If we have an indirection table already, we will get back a range in the indirection table here.
+    mdcursor_t existing_range;
+    uint32_t count;
+    if (!md_get_column_value_as_range(list_owner, list_col, &existing_range, &count))
+        return false;
+
+    if (CursorTable(&existing_range)->cxt == NULL)
+    {
+        // If we don't have a table to add the row to, create one.
+        if (!allocate_new_table(CursorTable(&list_owner)->cxt, CursorTable(&existing_range)->table_id))
+            return false;
+
+        // Now that we have a table, we recreate the "existing range" cursor as the one-past-the-end cursor
+        // This allows us to use the remaining logic unchanged.
+        existing_range = create_cursor(CursorTable(&existing_range), 1);
+    }
+
+    mdcursor_t row_after_range = existing_range;
+    // Move the cursor just past the end of the range. We'll insert a row at the end of the range.
+    if (!md_cursor_move(&row_after_range, count))
+        return false;
+    
+    return add_new_row_to_list(list_owner, list_col, row_after_range, new_row);
+}
+
+bool md_add_new_row_to_sorted_list(mdcursor_t list_owner, col_index_t list_col, col_index_t sort_order_col, uint32_t sort_col_value, mdcursor_t* new_row)
+{
+    if (!col_points_to_list(&list_owner, list_col))
+       return false;
+
+    // Get the range of rows already in the parent's child list.
+    // If we have an indirection table already, we will get back a range in the indirection table here.
+    mdcursor_t existing_range;
+    uint32_t count;
+    if (!md_get_column_value_as_range(list_owner, list_col, &existing_range, &count))
+        return false;
+
+    if (CursorTable(&existing_range)->cxt == NULL)
+    {
+        // If we don't have a table to add the row to, create one.
+        if (!allocate_new_table(CursorTable(&list_owner)->cxt, CursorTable(&existing_range)->table_id))
+            return false;
+
+        // Now that we have a table, we recreate the "existing range" cursor as the one-past-the-end cursor
+        // This allows us to use the remaining logic unchanged.
+        existing_range = create_cursor(CursorTable(&existing_range), 1);
+    }
+
+    mdcursor_t row_to_insert_before = existing_range;
+    // Move the cursor to just past the end of the range. If we don't find a place in the middle that we need to insert the row,
+    // we'll insert it here.
+    if (!md_cursor_move(&row_to_insert_before, count))
+        return false;
+
+    // In most cases we will be inserting at the end of the list,
+    // so start searching there to make this a little faster.
+    for (uint32_t i = count; i > 0; --i)
+    {
+        mdcursor_t row_to_check = existing_range;
+        if (!md_cursor_move(&row_to_check, i - 1))
+            return false;
```

Comment:
Can't we just compute the end and walk back until we get to `existing_range`? I would imagine updating this so it basically becomes `md_cursor_move(..., -1)` and `!= existing_range`?

I don't know if we have the APIs to do the above, but it might be worth considering that approach, which is more iterator based.

Reply:
We don't have a public mechanism for equality of cursors here other than `memcmp`, but internally I can use `CursorRow` to implement the algorithm similarly. That might be something we want to consider adding in the `dnmd.hpp` header for our C++ users.